src/lib/logical_query_plan/abstract_lqp_node.cpp

```diff
@@ -68,6 +69,33 @@ AbstractLQPNode::~AbstractLQPNode() {
   if (_inputs[1]) _inputs[1]->_remove_output_pointer(*this);
 }
 
+size_t AbstractLQPNode::hash() const {
+  size_t hash{0};
+
+  visit_lqp(std::const_pointer_cast<AbstractLQPNode>(shared_from_this()), [&hash](const auto& node) {
+    if (node) {
+      for (const auto& expression : node->node_expressions) {
+        boost::hash_combine(hash, expression->hash());
+      }
+      boost::hash_combine(hash, node->type);
+      boost::hash_combine(hash, node->_shallow_hash());
+      // since visit_lqp is used, the hash for an already visited node is not combined with
+      // the overall hash again, even if the same node is used as left and right input node (diamond structure).
+      // Therefore, a node that has only one (left) input node could have the same hash as a node with two
+      // (left and right) inputs which are in a diamond structure (same node pointer).
+      // To differentiate these, the boolean value (left input == right input) is combined with the overall hash.
+      boost::hash_combine(hash, node->left_input() == node->right_input());
```

Comment:
Maybe I missed the discussion around this, but when would left_input == right_input ever be non-false? In my mind: never.

Reply:
I remember understanding this a while back. :/

Reply:
I think I understand the *problem* that's trying to be addressed here: the hash is just comprised of the nodes, not of the graph structure. But this is, unless I am being stupid, not the solution. I have never seen a plan where the left input equals the right input and even if, that's not what we usually refer to when we say "diamond structure"

Reply:
Absolutely correct. Should it be `node->right_input() == nulltptr`?

Reply:
Don't think that would help. Whether inputs are set should be sufficiently encoded in the node type.

Reply:
As clarified in #1836, this line was only added for an UpdateNode-Test with only one (left) input node.